### PR TITLE
#4750 Java invoke local support for handlers that implement RequestStreamHandler

### LINKED ISSUE #4750: Java invoke local support for handlers that implement RequestStreamHandler
# This is a Feature Proposal

## Description

For bug reports:
* What went wrong?
* What did you expect should have happened?
* What was the config you used?
* What stacktrace or error message from your provider did you see?

For feature proposals:

When using Java to write a Lambda function, the handler can implement the RequestStreamHandler interface (see https://docs.aws.amazon.com/lambda/latest/dg/java-handler-using-predefined-interfaces.html and https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-create-api-as-simple-proxy-for-lambda.html#api-gateway-proxy-integration-lambda-function-java).

However, invoke local does not work in this case. It only currently works if the handler implements RequestHandler or does not use any of the predefined interfaces. 

This feature proposal is to add invoke local support for RequestStreamHandler.

Similar or dependent issues:
* None

## Additional Data

* ***Serverless Framework Version you're using***: 1.26.0
* ***Operating System***: Windows 10
* ***Stack Trace***: 
* ***Provider Error messages***:

### PR DESCRIPTION
**Pull request by XaeroDegreaz**
## What did you implement:

Closes #4750 

## How did you implement it:
I performed some checks inside the invoke method to support `RequestStreamHandler` compatible handlers. The checks include testing whether or not the first parameter was an `InputStream` type, and also a similar check during the reflected method invocation.

## How can we verify it:
Unit tests are included that verify that `RequestStreamHandler` methods can receive input, and write to an `OutputStream`. You can verify it by running the tests from your IDE, or simply run `mvn clean test` in the root of the plugin's directory.

You can also run `mvn clean package` from the plugin's directory; it will run the tests, and output the plugin JAR to `target/invoke-bridge-1.0.jar`. You can take this JAR file and put it in your global `node_modules` directory that houses `serverless`. For instance on my machine that location is `/usr/local/lib/node_modules/serverless/lib/plugins/aws/invokeLocal/java/target`. Be sure to back up the original `serverless` JAR, just in case ;)

After replacing your JAR, you can run `sls invoke local` with your function (one that handles `RequestStreamHandler`), and provide a path to some target data file to verify functionality.

## Todos:

- [x] Write tests
- [ ] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/plugins/aws/invokeLocal/java/src/test/resources/test.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      "Accept": "*/*",
+      "Accept-Encoding": "gzip, deflate",
+      "cache-control": "no-cache",
+      "CloudFront-Forwarded-Proto": "https",
+      "CloudFront-Is-Desktop-Viewer": "true",
+      "CloudFront-Is-Mobile-Viewer": "false",
+      "CloudFront-Is-SmartTV-Viewer": "false",
+      "CloudFront-Is-Tablet-Viewer": "false",
+      "CloudFront-Viewer-Country": "KR",
+      "Host": "rr4lbbmtfb.execute-api.us-east-1.amazonaws.com",
```
Comment (XaeroDegreaz) [inline lib/plugins/aws/invokeLocal/java/src/test/resources/test.json]: Old test dev APP id -- no longer valid.

File: lib/plugins/aws/invokeLocal/java/src/test/resources/test.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      "Postman-Token": "60d4d786-60bf-4ba5-bcde-5a00ac688041",
+      "User-Agent": "PostmanRuntime/7.3.0",
+      "Via": "1.1 17b1095550a8ffe13061114dbfceeb81.cloudfront.net (CloudFront)",
+      "X-Amz-Cf-Id": "QK3FgYHC89HWK03RsV7L8lJv34WzlFDyHmS0pbbCYtz9WyUCiw7JCQ==",
```
Comment (XaeroDegreaz) [inline lib/plugins/aws/invokeLocal/java/src/test/resources/test.json]: Old test dev CF id -- no longer valid.

**Comment by XaeroDegreaz** [comment]
@pmuens 

Hm, did you put the `invoke-bridge-1.0.1` JAR in your plugins directory?

I'm not entirely sure how one would test it other than that. I just packaged the jar manually, dumped it in the plugin directory, updated the `index.js` file in the parent directory, and ran the invoke command.

Once I saw that worked, I updated the `index.js` file in the repository and pushed.

How are you testing?

**Comment by pmuens** [comment]
Thanks for the reply @XaeroDegreaz 👍

I started with a regular `invoke local` test to check whether the core functionality (prior to this PR) still works.

So what I did was:

1. `serverless create -t aws-java-maven -p java-maven`
1. `cd java-maven`
1. `mvn package`
1. `serverless invoke local -f hello`

This works on `master`, but I got the permission error when using this PR.

**Comment by XaeroDegreaz** [comment]
@pmuens 

Ahh, I see!

Ok, so after `mvn package`, you need to take the artifact, `./target/invoke-bridge-1.0.1.jar`, and put it in `~/code/serverless/lib/plugins/aws/invokeLocal/java/target/`.

I'm assuming that you guys have some build process that does this for the **real** distribution that people get, i.e. performs the Maven build then puts the artifact where it needs to go for dist, however in this instance, you're simply checking out the repository as-is, building the JAR, and not doing anything with it :D

Hope it helps!

**Comment by pmuens** [comment]
Thanks for the follow-up @XaeroDegreaz 👍

I just tested it again and it works. The reason why it failed on my end was that I had the old bridge built (from `master`) which was not overwritten when running your changes. I just wiped the `target` directory in the `invokeLocal` plugin and everything worked as expected.

**Pull request merged**